### PR TITLE
[SYCL][Graph] in-order queue barrier fix

### PR DESCRIPTION
**Pull request by EwanC**
Fix for https://github.com/intel/llvm/issues/13066

The special case for using barriers on an in-order queue is that the last event/node submitted to the queue is used as an event for the barrier to depend on.

Looking at the last command submitted to the queue isn't correct for a graph, because previous commands
submitted to a graph could have been added explicitly or from recording another queue. Therefore, there is not guaranteed that the last command submitted by the in-order queue is correct dependency for the barrier node in the graph.

### CODE REVIEW

**Comment by EwanC** [comment]
The `InorderQueue/in_order_ext_oneapi_submit_barrier.cpp`  E2E test which is failing on HIP does have graph extension code in it https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/InorderQueue/in_order_ext_oneapi_submit_barrier.cpp#L72 So this is a fail that I need to look into (although it isn't yet clear to me why it only affects HIP backends)

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 1]
```diff
@@ -1703,11 +1703,15 @@ passed an invalid event.
...
 link:../supported/sycl_ext_oneapi_enqueue_barrier.asciidoc[sycl_ext_oneapi_enqueue_barrier]
 can only be used in graph nodes created using the Record & Replay API, as
-barriers rely on events to enforce dependencies. A synchronous exception will be
-thrown with error code `invalid` if a user tries to add them to a graph using
-the Explicit API. Empty nodes created with the `node::depends_on_all_leaves`
-property can be used instead of barriers when a user is building a graph with
-the explicit API.
+barriers rely on events to enforce dependencies. For barriers with an empty
+wait list parameter, the semantics are that the barrier node being added to
+will depend on all the existing graph leave nodes, not only the leave nodes
```
Comment (fabiomestre) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc:1708]: I find barriers affecting nodes from other queues a bit counter-intuitive. I'm wondering what is the reasoning behind this change? Is it required to fix the bug mentioned in https://github.com/intel/llvm/issues/13066 or are we just clarifying that that's how the implementation works at the moment?

Comment (EwanC) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc:1708]: Good question, this new wording is just clarifying the current implementation and is not related to the bugfix. However, when I was debugging the bugfix these semantics weren't immediately obvious so I thought best to state explicitly what's going on.

Comment (ori-sky) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc:1708]: Minor point but this should read as:
```suggestion
will depend on all the existing graph leaf nodes, not only the leaf nodes
```

**Review by ori-sky — CHANGES_REQUESTED** [review]
Looks good to me aside from the grammatical correction.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc (hunk 1 above)
Comment (fabiomestre) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc:1708]: Fair enough. I think that for this PR that sounds good. We could discuss the reasoning for the current implementation somewhere else.

Comment (EwanC) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: committed this change after Ori rightly informed by that the singular is "leaf" and plural is "leaves", so it's only my current text that's incorrect and should be updated.

**Review by cperkinsintel — APPROVED** [review]

**Review by fabiomestre — APPROVED** [review]

**Comment by EwanC** [comment]
@intel/llvm-gatekeepers Can you merge this please

**Pull request merged**